lib/statsd/instrument/udp_sink.rb

```diff
@@ -8,55 +11,74 @@ def self.for_addr(addr)
     new(host, Integer(port_as_string))
   end
 
+  class Dispatcher
+    def initialize(host, port, queue)
+      @host = host
+      @port = port
+      @queue = queue
+      @socket = nil
+      @thread = nil
+      ObjectSpace.define_finalizer(self, Proc.new { |o| o.stop })
```

Comment:
For my own education, when would you choose to use `at_exit` and when `ObjectSpace.define_finalizer`?

Reply:
`at_exit` is called when the process exists.

`define_finalizer` is called when the object is garbaged collected.

Here a finalizer is better, because if you dereference a sink, you want it to be cleaned up immediately rather than on the process exit.